server/worker/worker.go

```diff
@@ -0,0 +1,110 @@
+package worker
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+const maxRetries = 5
+
+// Job defines an interface for jobs that can be run by the Worker
+type Job interface {
+	// Name is the unique name of the job.
+	Name() string
+
+	// Run performs the actual work.
+	Run(ctx context.Context, args map[string]interface{}) error
+}
+
+// Worker runs jobs. NOT SAFE FOR CONCURRENT USE.
+type Worker struct {
+	ds  fleet.Datastore
+	log kitlog.Logger
+
+	registry map[string]Job
+}
+
+func NewWorker(ds fleet.Datastore, log kitlog.Logger) *Worker {
+	return &Worker{
+		ds:       ds,
+		log:      log,
+		registry: make(map[string]Job),
+	}
+}
+
+func (w *Worker) Register(jobs ...Job) {
+	for _, j := range jobs {
+		w.registry[j.Name()] = j
+	}
+}
+
+func (w *Worker) QueueJob(ctx context.Context, job *fleet.Job) (*fleet.Job, error) {
+	job.State = fleet.JobStateQueued
+	return w.ds.NewJob(ctx, job)
+}
+
+// ProcessJobs processes all queued jobs.
+func (w *Worker) ProcessJobs(ctx context.Context) error {
+
+	// process jobs until there are none left
```

Comment:
We could run into situations where the queue grows faster than the processing. It should run for as long as it has the lock, but not (too much) beyond. Ideally, this would be controlled with the context, and the loop would check if the context is done before starting processing another job (and that processing would do a best effort to stop when the context is cancelled too).